### PR TITLE
This fixes several items, see commit comments

### PR DESCRIPTION
**Pull request by Levi-Armstrong**

### CODE REVIEW

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (marip8) [inline tesseract_command_language/include/tesseract_command_language/joint_waypoint.h:293]: Why don't you use `Eigen::isApprox(upper_tolerance.cwiseAbs(), lower_tolerance.cwiseAbs())`? The `1e-5` seems arbitrary; if you care about them being the same you should use machine `epsilon` which is think is the default for `Eigen::isApprox`

File: README.md
Comment (marip8) [inline README.md:52]:
```suggestion
### `tcmalloc` (recommended)
```

File: README.md
Comment (marip8) [inline README.md:54]:
```suggestion
The recommended approach to address memory management is to leverage the `tcmalloc` library which is part of Google Performance Tools (`gperftools`). Experimentation has shown `tcmalloc` to provide more stable memory management when leveraging the planning server.
```

File: README.md
Comment (marip8) [inline README.md:56]:
```suggestion
This capability can be leveraged by linking the exectuable that contains the planning server against `tcmalloc` using the following commands:
```

File: README.md
Comment (marip8) [inline README.md:76]:
```suggestion
### `ptmalloc` (Ubuntu OS default)
```

File: README.md
Comment (marip8) [inline README.md:78]:
```suggestion
`ptmalloc` provides run-time [tunables](https://www.gnu.org/software/libc/manual/html_node/Tunables.html) through a set of environment variables.
```

File: README.md
Comment (marip8) [inline README.md:80]: You should elaborate on why giving memory back to the system is important and why it doesn't do it by default.

Also `tunnable` -> `tunable`

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h:293]: > Eigen::isApprox(upper_tolerance.cwiseAbs(), lower_tolerance.cwiseAbs())

We probably don't want to use cwiseAbs() because -pi would show equal to pi, but I see where you are going.

Suggest changing to: `Eigen::isApprox(upper_tolerance, lower_tolerance)`?

File: README.md
Comment (Levi-Armstrong) [reply README.md:80]: I believe the main paragraph under `Process Planning Server Tuning` explains this, does it not?

See below:

The planning server is built leveraging the library taskflow which enables you to quickly write parallel and heterogeneous tasks programs which aligns well with process motion planning. These types of libraries leverage thread pools to quickly spin up different tasks as needed. One interesting thing found, if a task consumes a lot of memory libc will make a performance decision to allow the thread to hold onto the memory it has for subsequent tasks opposed to giving the memory back to your system. This may be fine if this was the only set of tasks being run but typically in our use case there are other computationally intensive processes which could use this memory.

Comment (marip8) [reply README.md:80]: Okay I think that's fine. I must have skipped over that section

**Comment by Levi-Armstrong** [comment]
@marip8 I made the requested changes but have two open items left to discuss. I will clean up the commit history once approved.

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (marip8) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: I was thinking that `lower_tolerance` might be negative and need to have the absolute value taken. If both vectors are positive then that would work

File: README.md
Comment (Levi-Armstrong) [reply README.md:80]: Yea it most likely was not in the changes because I don't think I changed this paragraph.

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (marip8) [inline tesseract_command_language/include/tesseract_command_language/joint_waypoint.h:291]: Is this returning the opposite of what it did before? It seems like previously true meant that the values were different (i.e.  `sum` was larger than the tolerance) and now true means that they are the same

Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h:291]: Yep good catch it should be `!Eigen::IsApprox(lower_tolerance, upper_tolerance);`

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: I just assume that `Eigen::isApprox` handles this when checking against the epsilon but not sure if that is true.

Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: It looks like the isApprox function should handle this; do you agree?

The fuzzy compares are done multiplicatively. Two vectors v and w are considered to be approximately equal within precision p if

∥v−w∥⩽pmin(∥v∥,∥w∥).
For matrices, the comparison is done using the Hilbert-Schmidt norm (aka Frobenius norm L2 norm).
Because of the multiplicativeness of this comparison, one can't use this function to check whether *this is approximately equal to the zero matrix or vector. Indeed, isApprox(zero) returns false unless *this itself is exactly the zero matrix or vector. If you want to test whether *this is zero, use internal::isMuchSmallerThan(const RealScalar&, RealScalar) instead.

Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: Now that I read the description I am not sure we should use this check by itself because of the zero case.

Comment (marip8) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: It appears that `isApprox` doesn't do element-wise comparison, and it also doesn't make sure both vectors are positive. [Here is the description](https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#ae8443357b808cd393be1b51974213f9c)

Comment (marip8) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: I don't think this will work if one of the vectors is negative because of the subtraction on the left hand side. If you're going to use this, the vectors both need to be positive. Also if they can be zero vectors (which I assume they are by default), then the original change might be better (with the addition of a comparison to machine epsilon instead of 1e-5)

Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: This approach `Eigen::isApprox(upper_tolerance.cwiseAbs(), lower_tolerance.cwiseAbs())` fails in the unit test. The following `!Eigen::IsApprox(lower_tolerance, upper_tolerance)` works but based on the description it does pose concern for the all zero case. I created a vectorized version of `almostEqualRelativeAndAbs` which is probably the best solution.

``` c++
  std::vector<std::string> joint_names = {"joint_1", "joint_2", "joint_3"};
  Eigen::VectorXd joint_values = Eigen::VectorXd::Constant(3, 1);

  JointWaypoint jw(joint_names, joint_values);
  EXPECT_FALSE(jw.isToleranced());

  jw.upper_tolerance = Eigen::VectorXd::Constant(3, 5);
  jw.lower_tolerance = Eigen::VectorXd::Constant(3, -5);
  EXPECT_TRUE(jw.isToleranced());

  jw.upper_tolerance = Eigen::VectorXd::Constant(3, 0);
  jw.lower_tolerance = Eigen::VectorXd::Constant(3, 0);
  EXPECT_FALSE(jw.isToleranced());
```

Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: I update the PR to leverage the `almostEqualRelativeAndAbs` function.

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: updated to leverage `almostEqualRelativeAndAbs`

**Comment by Levi-Armstrong** [comment]
This PR depend on this tesseract [PR](https://github.com/ros-industrial-consortium/tesseract/pull/536).

**Comment by Levi-Armstrong** [comment]
@marip8 and @mpowelson This PR good to merge?

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (mpowelson) [inline tesseract_command_language/include/tesseract_command_language/joint_waypoint.h:292]: The same code is present in cartesian_waypoint.

**Review by mpowelson — APPROVED** [review]

File: tesseract_command_language/test/joint_waypoint_test.cpp
Comment (marip8) [inline tesseract_command_language/test/joint_waypoint_test.cpp:43]: I think this technically violates your description of the tolerances:

>   /** Joint distance below waypoint that is allowed. **Each element should be <= 0** */
  Eigen::VectorXd lower_tolerance;

**Comment by marip8** [comment]
The descriptions of upper and lower tolerances seem a bit ambiguous to me. I think the way that you are intending to use them is similar to mechanical drawings where tolerances can be positive or negative and they are always added to the nominal value to get the toleranced value. Here are some simple examples, each of which is valid

```
5 + [-0.1, +0.1] = [4.9, 5.1] -> is toleranced
5 + [+0.1, +0.2] = [5.1, 5.1] -> is toleranced
5 + [-0.2, -0.1] = [4.8, 4.9] -> is toleranced
5 + [+0.1, +0.1] = [5.1, 5.1] -> is **not** toleranced
```

I think the checks you need to make are:
1. `lower_tolerance` is always element-wise smaller than `upper_tolerance`
2. If `lower_tolerance` is equal to `upper_tolerance

I would also suggest updating the documentation to make these assumptions about the tolerances clear

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (marip8) [inline tesseract_command_language/include/tesseract_command_language/joint_waypoint.h:292]:
```suggestion
    // Check if the upper tolerance is greater than the lower tolerance by some small amount
    return ((upper_tolerance - lower_tolerance).array() > max_diff).all();
```

File: tesseract_command_language/test/joint_waypoint_test.cpp
Comment (Levi-Armstrong) [reply tesseract_command_language/test/joint_waypoint_test.cpp:43]: I am not following how this violate the description; is there something I am missing? The lower tolerance is being populated with a vector of length 3 with a constant value of `-5`. All values in the vector should be `-5` which is `<=0`.

Comment (marip8) [reply tesseract_command_language/test/joint_waypoint_test.cpp:43]: Sorry, I misread this. Disregard my previous comment

File: tesseract_command_language/include/tesseract_command_language/joint_waypoint.h
Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: added

Comment (Levi-Armstrong) [reply tesseract_command_language/include/tesseract_command_language/joint_waypoint.h]: @marip8 The changes discussed in our meeting have been applied. Let me know if you see an issues with the changes.

**Review by marip8 — APPROVED** [review]

**Comment by marip8** [comment]
Just to summarize our earlier discussion:
- Lower tolerance is a set of numbers that should be element-wise <= 0.0
- Upper tolerance is a set of numbers that should be element-wise >= 0.0
- The joint state limit is calculated by adding the tolerance to the nominal value

File: tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp
Comment (Levi-Armstrong) [inline tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:287]: The paper was multiplying by 0.5.

File: tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp
Comment (Levi-Armstrong) [inline tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:866]: This change seems to make the trajectory duration increase for the tests. It was dividing by very small numbers because they never actually equaled zero.

File: tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp
Comment (Levi-Armstrong) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:287]: Also this change makes the trajectory duration increase. Not sure why it is different. 

From the [paper](http://www.roboticsproceedings.org/rss08/p27.pdf):

>where  the  first  two  elements  give  the  maximum  possible distances such that the circular segment does not replace more than half of the adjoining linear segments and the last element limits  the  radius  to  make  sure  the  circlular  segment  stays within a distance δ from waypoint qi

Comment (Levi-Armstrong) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:287]: @mpowelson let me know if you have an issue with this change and I can change it back.

Comment (mpowelson) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:287]: It's possible that these changes were added by the move it folks. I'm not sure how closely this follows the original implementation.

Comment (gavanderhoorn) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:287]: The PR which added it to MoveIt was pretty well structured. See https://github.com/ros-planning/moveit/pull/1365.

The first commit imports the upstream code, directly from the author of the paper. Later commits add the wrappers and MoveIt adapters.

Comment (Levi-Armstrong) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp]: @mpowelson The original commit form the author had it without the 0.5 factor so I have reverted my change to keep it the same.

@gavanderhoorn Thank you for the information and I was able to found the commit which added the change `acos(std::max(-1,...` so I added the comments from the PR in the code for future readers if they compare back to the paper.

Comment (gavanderhoorn) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp]: I would perhaps even just straight link to the MoveIt PRs.

And if we fix something here, contribute it back.

File: tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp
Comment (gavanderhoorn) [inline tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:831]:
```suggestion
    // It is possible to have both slope and start_slope to be equal
```

File: tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp
Comment (gavanderhoorn) [inline tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:838]: this could perhaps do with a comment? What happens if they are equal? And why? What if they are not?

Comment (Levi-Armstrong) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:838]: > What happens if they are equal? 

This occurs if two consecutive `TrajectorySteps` have the same acceleration. Resulting in `intersection_path_pos` being nan because it divides by zero.

> What if they are not?

I have not looked at the paper to see what is being calculated here.

Comment (gavanderhoorn) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp:838]: I (perhaps wrongfully) assumed this was added, due to the `+` in the gutter.

File: tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp
Comment (Levi-Armstrong) [reply tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp]: > I would perhaps even just straight link to the MoveIt PRs.

Done.

> And if we fix something here, contribute it back.

Ideally that would be preferred, though I personally don't have the bandwidth for this. Though my preference would be to create a separate repository for time parameterization and join resource with maintaining it since these are independent of Tesseract and MoveIt. Not sure if the maintainers of MoveIt would be open to this, but I would be.

**Pull request merged**
